### ISSUE TITLE
Package vs plugin vs feature

### ISSUE THREAD

**Issue by Reinmar** [ckeditor/ckeditor5-design#92]
## The future

As a part of https://github.com/ckeditor/ckeditor5/issues/45 I'm going to change a bit the architecture of packages and plugins. A package (in NPM sense) can contain 0-N modules. Some of these modules (0-N) can represent pluggable components / end-features (for now called "plugins").

A tool library would usually be a package with 0 plugins (it can contain modules which are not plugins). A basicstyle package will contain "bold", "italic", etc. plugins plus perhaps some other modules (like "basestyle" helper class). Another interesting example, of composable feature would be an "image" package with "imagebase", "imagecaption", "imagealignment", "imageupload", etc. features.

Another thing will be that you could theoretically have a package containing a skin, creator and a couple of features. Simply, some modules of this package would be a skin classes, some a creator and some a features. I think that we should recommend to not create such bundles or even disallow this, but the architecture we'll have would by default allow this.

All this means that we can imagine the addons (another name :D) repo for CKEditor 5 as a list of packages, each containing 0-N features/skins/creators which can be enabled. If we'd restrict the contents of packages to a single type at once (either a skin, or a creator or a set of features), then the addons repo would be a set of skins (from which you would be picking one), plus a set of creators (from which you would be picking one) and a set of packages with features (some packages with 0 features). It's all up to us still how we'll define the different kind of addons (not to be discussed in this ticket – I just wanted to paint some picture of how it will look).

## The past

In CKEditor 4 the situation was different. We had pieces of code called which were representing skins. And pieces of code which were representing plugins. There were no modules at all. The packages equaled plugins and you couldn't have a package containing multiple plugins.

## Naming problem

I've got two ideas for the naming conventions and I'm unsure which is better.

1. **Package** == **Plugin**. A **Package** is a set of **Modules**. Some **Modules** can be **Features** (or **Skins** or **Creators**).
  * Comment: we use the word "feature" in exactly this sense a lot. We've been actually discussing the "feature API". The word "plugin" is more vague IMO and covers more than just end-features, so it well covers also skins and creators.
  * Doubts: I'm not sure whether we should totally stop using the word "plugin", because, at least for me, plugins are the atomic entities. You plug in the entire plugin. Also, I think it could be odd that a plugin can contain a skin.
2. Like "1", but we totally drop the name **Plugin**.
3. A **Package** is a set of **Modules**. Some **Modules** can be **Plugins** (or **Skins** or **Creators**).
  * This scenario keeps the name "plugin" but we group them in packages.
  * I personally don't like this proposal because for years we've been talking about "features". I think that's because they better reflect what they really are. On the other hand, it may be better understandable for outsiders.

## Questions

I think the question is clear – which naming convention makes most sense and why.

**Comment by oleq** [comment]
My perception of **Plugin** (as a term) tells me that it brings something extra to the application. It could be a **Skin** (**Theme**!), **Feature** or some **Creator**, it doesn't really matter. So **Package == Plugin** is a valid assumption for me.

Then, considering that **Package == Plugin**, the conclusions is that **Plugin != Feature**. It leaves options 1. and 2.

Then, there's the social aspect. The **Plugin**–oriented architecture is a preferred way of building software and people got used to this approach. But the problem is that we must deliver **Themes** in the same way as **Features** or **Creators** while the common practice (just take a look on [Wordpress](https://wordpress.org/themes/)) is that a live code (**Features** or **Creators**) is shipped separately from **Themes** (static resources). And I think it is a correct assumption that people might think that way when it comes to integrating CKE5 and searching our websites for some extras. This is also why I would **avoid the *Plugin* word, despite its social meaning and importance**: because it could be very confusing for some developers that **Themes** are encapsulated in **Plugins**.

So, long story short, **I'm for 2**. And we're not alone because there's a software which uses similar approach: Sublime Text. They deliver both **Features** and **Themes** in **Packages** and avoid the **P**–word. It makes sense to me.

**Comment by wwalc** [comment]
+1 for option 2 because of the reasons mentioned by oleq. Considering that the role of a "plugin" in V5 will be changed comparing to previous versions of CKEditor using this name would be confusing.

**Comment by pjasiun** [comment]
+1 for option 2. It is the simples one.

**Comment by AnnaTomanek** [comment]
This must be the magic of Christmas or something - I also agree, +1 for option 2, for the reasons explained above.

**Comment by Reinmar** [comment]
Thank you for your opinions. I preferred option 2 as well, so we've got 5:0 :).

**Comment by Reinmar** [comment]
I've finished updating docs with the recent updates. One of them was the redefinition of plugins. We kept them, because while implementing features and creators I understood that despite having a different purpose, they need to be loaded in the same way and have some similar features like async initialization and the ability to require other features.

You can read more in https://github.com/ckeditor/ckeditor5-design/wiki/Architecture-Overview#plugins and https://github.com/ckeditor/ckeditor5-design/wiki/Plugins and https://github.com/ckeditor/ckeditor5-design/wiki/Packages
